bd-client-stats/src/file_manager.rs

```diff
@@ -0,0 +1,424 @@
+  async fn get_initialized(&mut self) -> anyhow::Result<&mut InitializedInner> {
+    // Due to the way logger startup works we delay initializing the index until first use. First we
+    // try to load an existing file index. If this doesn't exist or is corrupted we remove the
+    // entire directory as there is no reasonable way to manage the contents.
+    match self {
+      Self::Initialized(inner) => Ok(inner),
+      Self::NotInitialized(file_system) => {
+        let file_system = file_system.take().unwrap();
+        let path = STATS_DIRECTORY.join(&*PENDING_AGGREGATION_INDEX_FILE);
+        log::debug!("initializing pending aggregation index: {}", path.display());
+        let index = match file_system
+          .read_file(&path)
+          .await
+          .and_then(|contents| read_compressed_protobuf(&contents))
+        {
+          Ok(index) => index,
+          Err(e) => {
+            log::debug!("unable to open pending aggregation index: {e}");
+            log::debug!("creating new aggregation index");
+
+            file_system.remove_dir(&STATS_DIRECTORY).await?;
+            file_system.create_dir(&STATS_DIRECTORY).await?;
+            PendingAggregationIndex::default()
+          },
+        };
+
+        *self = Self::Initialized(InitializedInner {
+          file_system,
+          index: index.pending_files.into(),
+        });
+
+        Ok(match self {
+          Self::Initialized(inner) => inner,
+          Self::NotInitialized(_) => unreachable!(),
+        })
+      },
+    }
+  }
+}
+
+impl FileManager {
+  pub fn new(
+    file_system: Box<dyn FileSystem>,
+    time_provider: Arc<dyn TimeProvider>,
+    runtime_loader: &ConfigLoader,
+  ) -> anyhow::Result<Self> {
+    Ok(Self {
+      inner: Mutex::new(Inner::NotInitialized(Some(file_system))),
+      time_provider,
+      max_aggregated_files: runtime_loader.register_watch()?,
+      max_aggregation_window_per_file: runtime_loader.register_watch()?,
+    })
+  }
+
+  // Read an existing snapshot from disk to merge into, or create a new one.
+  pub async fn get_or_create_snapshot(&self) -> anyhow::Result<StatsUploadRequestHandle> {
+    let mut inner = self.inner.lock().await;
+    let initialized_inner = inner.get_initialized().await?;
+
+    let use_existing = if initialized_inner.index.back().map_or(true, |file| {
+      if file.ready_to_upload {
+        log::debug!("snapshot is ready to upload, creating new snapshot");
+        true
+      } else if file.period_start.to_offset_date_time()
+        + self.max_aggregation_window_per_file.read()
+        <= self.time_provider.now()
+      {
+        log::debug!("snapshot is too old, creating new snapshot");
+        true
+      } else {
+        false
+      }
+    }) {
+      if self.max_aggregated_files.read() <= u32::try_from(initialized_inner.index.len()).unwrap() {
+        log::debug!("max files reached, popping oldest snapshot");
+        initialized_inner.delete_oldest_snapshot().await?;
```

Comment:
obviously a bit hard to emit stats within the stats system but would be nice to have a way to track how often we end up expiring snapshots

Reply:
Yeah not sure how to do this. I guess I can add an internal stat but it might not be very accurate.